src/natcap/invest/seasonal_water_yield/seasonal_water_yield.py

```diff
@@ -879,133 +875,178 @@ def _calculate_annual_qfi(qfm_path_list, target_qf_path):
+        # precip mask: both p_im and n_m are defined and greater than 0
+        precip_mask = valid_p_mask & valid_n_mask & (p_im > 0) & (n_m > 0)
+        stream_mask = stream == 1
+        # stream_nodata is the only input that carries over nodata values from
         # the aligned DEM.
-        if stream_nodata is not None:
-            valid_mask &= ~utils.array_equals_nodata(
-                stream_array, stream_nodata)
-
-        valid_n_events = n_events[valid_mask]
-        valid_si = s_i[valid_mask]
+        valid_mask = (
+          valid_p_mask &
+          valid_n_mask &
+          ~utils.array_equals_nodata(stream, stream_nodata) &
+          ~utils.array_equals_nodata(s_i, si_nodata))
+
+        # QF is defined in terms of three cases:
+        #
+        # 1. Where there is no precipitation, QF = 0
+        #    (even if stream or s_i is undefined)
+        #
+        # 2. Where there is precipitation and we're on a stream, QF = P
+        #    (even if s_i is undefined)
+        #
+        # 3. Where there is precipitation and we're not on a stream, use the
+        #    quickflow equation (only if all four inputs are defined):
+        #    QF_im = 25.4 * n_m * (
+        #       (a_im - s_i) * exp(-0.2 * s_i / a_im) +
+        #       s_i^2 / a_im * exp(0.8 * s_i / a_im) * E1(s_i / a_im)
+        #    )
+        #
+        # When evaluating case 3, there are a couple of edge cases related
+        # to the last term of the quickflow equation:
+        #
+        #       s_i^2 / a_im * exp(0.8 * s_i / a_im) * E1(s_i / a_im)
+        #
+        # 3a. Where s_i = 0, this term should equal 0 (per conversation with
+        #     Rafa). But, evaluating in numpy you get NaN and a warning
+        #     because E1(0 / a_im) = infinity.
+        #
+        #     Solution: Preemptively set the result to 0 where s_i = 0 in
+        #     order to avoid calculations with infinity.
+        #
+        # 3b. When the ratio s_i / a_im becomes large, this term approaches 0.
+        #     [NOTE: I don't know how to prove this mathematically, but it
+        #     holds true when I tested with reasonable values of s_i and a_im].
+        #     The exp() term becomes very large, while the E1() term becomes
+        #     very small. When (s_i / a_im) > 880 ish, exp(0.8 * s_i / a_im)
+        #     will overflow and round up to infinity. Meanwhile, E1(s_i / a_im)
+        #     will become so small that it rounds down to 0.
+        #
+        #     Solution: Catch the overflow warning, and set the term to 0
+        #     anywhere that overflow happens.
+        #
+        # 3c. Otherwise, evaluate the term as usual.
+
+        # qf_im is the quickflow at pixel i on month m
+        qf_im = numpy.full(p_im.shape, TARGET_NODATA, dtype=numpy.float32)
+
+        # case 1: there is no precipitation where both p_im and n_m are
+        # defined and equal to zero.
+        case_1_mask = ~precip_mask
```

Comment:
because precip_mask is defined above as:
```
# precip mask: both p_im and n_m are defined and greater than 0
precip_mask = valid_p_mask & valid_n_mask & (p_im > 0) & (n_m > 0)
```
it could include pixels that have nodata in the `s_i` or `stream` arrays, but not in the precip or n_events arrays. 

That's because you don't need to know `s_i` or `stream` to calculate `QF = 0 when P = 0`. I think you could argue that it should be nodata anyway, but this is consistent with the past implementation.